README.md
[![Build Status](https://travis-ci.org/n3okill/enfsmove.svg)](https://travis-ci.org/n3okill/enfsmove)
[![Build status](https://ci.appveyor.com/api/projects/status/7sa7do8hf79c4j5d?svg=true)](https://ci.appveyor.com/project/n3okill/enfsmove)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/8721ee3980094b75913805efc931bff4)](https://www.codacy.com/app/n3okill/enfsmove)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfsmove.png)](https://nodei.co/npm/enfsmove/)

enfsmove
=========
Module that add move functionality to node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)


Description
-----------
This module will add a method that allows moving files and directories in the file system.

- This module will add following methods to node fs module:
  * move
  * moveSync
  
Usage
-----
`enfsmove`

```js
    var enfsmove = require("enfsmove");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [move](#move)
- [moveSync](#movesync)


### move
  - **move(srcPath, dstPatch, [options], callback)**

> Move files and directories in the file system

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * mkdirp (Boolean): if true will create new directories instead of copying the old ones (default: false)
  * overwrite (Boolean): if true will overwrite items at destination if they exist (Default: false)
  * limit (Number): the maximum number of items being moved at a moment (Default: 512)
  

```js
    enfsmove.move("/home/myHome","/home/myOtherHome", function(err){
        if(!err) {
            console.log("Everything moved correctly");
        }
    });
```


### moveSync
  - **moveSync(srcPath, dstPath, [options])**

> Move files and directories in the file system

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * mkdirp (Boolean): if true will create new directories instead of copying the old ones (default: false)
  * overwrite (Boolean): if true will overwrite items at destination if they exist (Default: false)
  * limit (Number): the maximum number of items being moved at a moment (Default: 512)


```js
    enfsmove.moveSync("/home/myHome","/home/myOtherHome");
    console.log("Everything moved correctly");
```


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-03-31
==================
  * code improvement
  * dependencies update
  * major version

0.1.0 / 2016-04-11
==================
  * es6 code update
  * dependencies update

0.0.1 / 2016-02-22
==================

  * first release


test/moveSync.js
/**
 * Created by JParreir on 05-10-2015.
 */
/* global afterEach, beforeEach, describe, it, after, before, process */

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const rimraf = require("rimraf");
const enFs = require("fs");
const enfsmkdirp = require("enfsmkdirp");
const move = require("../").moveSync;
const copy = require("enfscopy");
const cwd = process.cwd();

describe("enfsmoveSync", function () {
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfsmovesync");
    const helperPath = nodePath.join(__dirname, "helper");
    const tmpHelperPath = nodePath.join(tmpPath, "helper");
    before(function (done) {
        enfsmkdirp.mkdirp(tmpPath, function () {
            process.chdir(tmpPath);
            done();
        });
    });
    beforeEach(function (done) {
        copy.copy(helperPath, tmpHelperPath, done);
    });
    after(function (done) {
        rimraf(tmpPath, function () {
            process.chdir(cwd);
            done();
        });
    });
    afterEach(function (done) {
        rimraf(tmpPath + nodePath.sep + "*", done);
    });

    function createAsyncErrFn(errCode) {
        function fn() {
            fn.callCount++;
            const callback = arguments[arguments.length - 1];
            setImmediate(() => {
                const err = new Error();
                err.code = errCode;
                callback(err);
            });
        }

        fn.callCount = 0;
        return fn;
    }

    function createSyncErrFn(errCode) {
        function fn() {
            fn.callCount++;
            const err = new Error();
            err.code = errCode;
            throw err;
        }

        fn.callCount = 0;
        return fn;
    }

    const originalRename = enFs.rename;
    const originalLink = enFs.link;
    const originalRenameSync = enFs.renameSync;
    const originalLinkSync = enFs.linkSync;


    function setUpMockFs(errCode) {
        enFs.rename = createAsyncErrFn(errCode);
        enFs.link = createAsyncErrFn(errCode);
        enFs.renameSync = createSyncErrFn(errCode);
        enFs.linkSync = createSyncErrFn(errCode);
    }

    function tearDownMockFs() {
        enFs.rename = originalRename;
        enFs.link = originalLink;
        enFs.renameSync = originalRenameSync;
        enFs.linkSync = originalLinkSync;
    }

    it("should rename a file on the same device", function () {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpHelperPath, "file1-dst");
        move(src, dst);
        enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
    });
    it("should not overwrite if {overwrite=false}", function () {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpHelperPath, "file1-dst");
        enFs.writeFileSync(dst, "new content");
        (function () {
            move(src, dst, {overwrite: false});
        }).should.throw(Error, {code: "EEXIST"});
        enFs.readFileSync(dst, "utf8").should.be.equal("new content");
    });
    it("should not create directory structure if mkdirp is false", function () {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        (function () {
            move(src, dst, {mkdirp: false});
        }).should.throw(Error, {code: "ENOENT"});
    });
    it("should create directory structure by default", function () {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        move(src, dst);
        enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
    });
    it("should work across devices", function () {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        setUpMockFs("EXDEV");
        move(src, dst, {fs: enFs});
        enFs.linkSync.callCount.should.be.equal(1);
        enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
        tearDownMockFs("EXDEV");
    });
    it("should move folders", function () {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        move(src, dst);
        enFs.readFileSync(nodePath.join(dst, "file2"), "utf8").should.containEql("data file 2");
    });
    it("should move folders across devices with EISDIR error", function () {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        setUpMockFs("EISDIR");

        move(src, dst, {fs: enFs});
        enFs.linkSync.callCount.should.be.equal(1);
        enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
        tearDownMockFs("EISDIR");
    });
    it("should overwrite folders across devices", function () {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        enfsmkdirp.mkdirpSync(dst);
        setUpMockFs("EXDEV");
        move(src, dst, {overwrite: true, fs: enFs});
        enFs.renameSync.callCount.should.be.equal(1);
        enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
        tearDownMockFs("EXDEV");
    });
    it("should move folders across devices with EXDEV error", function () {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        setUpMockFs("EXDEV");
        move(src, dst, {fs: enFs});
        enFs.linkSync.callCount.should.be.equal(1);
        enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
        tearDownMockFs("EXDEV");
    });
    //describe("> when trying to a move a folder into itself", function() {
    it("should produce an error", function () {
        const src = nodePath.join(tmpPath, "test");
        const dst = nodePath.join(tmpPath, "test", "test");
        (function () {
            enFs.statSync(src);
        }).should.throw(Error);
        enfsmkdirp.mkdirpSync(src);
        (function () {
            move(src, dst);
        }).should.throw(Error);
        enFs.statSync(src).isDirectory().should.be.equal(true);
    });
    it("should overwrite the destination", function () {
        const src = nodePath.join(tmpPath, "src");
        const dst = tmpHelperPath;
        // use fixtures dir as dst since it has stuff
        const files = enFs.readdirSync(dst);
        files.indexOf("file1").should.be.greaterThanOrEqual(0);
        files.indexOf("folder1").should.be.greaterThanOrEqual(0);

        //create new src dir
        enfsmkdirp.mkdirpSync(nodePath.join(src, "folder-test"));
        enFs.writeFileSync(nodePath.join(src, "some-file"), "data some-file");
        const filesSrc = enFs.readdirSync(src);
        filesSrc.indexOf("some-file").should.be.greaterThanOrEqual(0);
        filesSrc.indexOf("folder-test").should.be.greaterThanOrEqual(0);
        move(src, dst, {overwrite: true, fs: enFs});
        const filesDst = enFs.readdirSync(dst);
        //dst should not have old stuff
        filesDst.indexOf("file1").should.be.equal(-1);
        filesDst.indexOf("folder1").should.be.equal(-1);
        //dst should have new stuff
        filesDst.indexOf("some-file").should.be.greaterThanOrEqual(0);
        filesDst.indexOf("folder-test").should.be.greaterThanOrEqual(0);
    });
});


test/moveAsync.js
/**
 * Created by JParreir on 05-10-2015.
 */
/* global afterEach, beforeEach, describe, it, after, before, process */

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const rimraf = require("rimraf");
const enFs = require("fs");
const enfsmkdirp = require("enfsmkdirp");
const move = require("../").move;
const copy = require("enfscopy");
const cwd = process.cwd();

describe("enfsmoveAsync", function () {
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfsmoveasync");
    const helperPath = nodePath.join(__dirname, "helper");
    const tmpHelperPath = nodePath.join(tmpPath, "helper");

    before(function (done) {
        enfsmkdirp.mkdirp(tmpPath, function () {
            process.chdir(tmpPath);
            done();
        });
    });
    beforeEach(function (done) {
        copy.copy(helperPath, tmpHelperPath, done);
    });
    after(function (done) {
        rimraf(tmpPath, function () {
            process.chdir(cwd);
            done();
        });
    });
    afterEach(function (done) {
        rimraf(tmpPath + nodePath.sep + "*", done);
    });

    function createAsyncErrFn(errCode) {
        function fn() {
            fn.callCount++;
            const callback = arguments[arguments.length - 1];
            setImmediate(() => {
                const err = new Error();
                err.code = errCode;
                callback(err);
            });
        }

        fn.callCount = 0;
        return fn;
    }

    const originalRename = enFs.rename;
    const originalLink = enFs.link;

    function setUpMockFs(errCode) {
        enFs.rename = createAsyncErrFn(errCode);
        enFs.link = createAsyncErrFn(errCode);
    }

    function tearDownMockFs() {
        enFs.rename = originalRename;
        enFs.link = originalLink;
    }


    it("should rename a file on the same device", function (done) {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpHelperPath, "file1-dst");

        move(src, dst, (err) => {
            (err === null).should.be.equal(true);
            enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
            done();
        });
    });
    it("should not overwrite if {overwrite=false}", function (done) {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpHelperPath, "file1-dst");
        enFs.writeFileSync(dst, "new content");
        move(src, dst, {overwrite: false}, (err) => {
            err.should.be.instanceOf(Error);
            err.code.should.be.equal("EEXIST");
            enFs.readFileSync(dst, "utf8").should.be.equal("new content");
            done();
        });
    });
    it("should not create directory structure if mkdirp is false", function (done) {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        move(src, dst, {mkdirp: false}, (err) => {
            err.should.be.instanceOf(Error);
            err.code.should.be.equal("ENOENT");
            done();
        });
    });
    it("should create directory structure by default", function (done) {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        move(src, dst, (err) => {
            (err === null).should.be.equal(true);
            enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
            done();
        });
    });
    it("should work across devices", function (done) {
        const src = nodePath.join(tmpHelperPath, "file1");
        const dst = nodePath.join(tmpPath, "folder", "does", "not", "exist", "file1-dst");
        setUpMockFs("EXDEV");
        move(src, dst, {fs: enFs}, (err) => {
            (err === null).should.be.equal(true);
            enFs.link.callCount.should.be.equal(1);
            enFs.readFileSync(dst, "utf8").should.containEql("data file 1");
            tearDownMockFs("EXDEV");
            done();
        });
    });
    it("should move folders", function (done) {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");
        (function () {
            enFs.statSync(dst);
        }).should.throw(Error);
        move(src, dst, (err) => {
            (err === null).should.be.equal(true);
            enFs.readFileSync(nodePath.join(dst, "file2"), "utf8").should.containEql("data file 2");
            done();
        });
    });
    it("should move folders across devices with EISDIR error", function (done) {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        setUpMockFs("EISDIR");

        move(src, dst, {fs: enFs}, (err) => {
            (err === null).should.be.equal(true);
            enFs.link.callCount.should.be.equal(1);
            enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
            tearDownMockFs("EISDIR");
            done();
        });
    });
    it("should overwrite folders across devices", function (done) {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        enfsmkdirp.mkdirpSync(dst);
        setUpMockFs("EXDEV");
        move(src, dst, {fs: enFs, overwrite: true}, (err) => {
            (err === null).should.be.equal(true);
            enFs.rename.callCount.should.be.equal(1);
            enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
            tearDownMockFs("EXDEV");
            done();
        });
    });
    it("should move folders across devices with EXDEV error", function (done) {
        const src = nodePath.join(tmpHelperPath, "folder1");
        const dst = nodePath.join(tmpPath, "folder1-dst");

        setUpMockFs("EXDEV");
        move(src, dst, {fs: enFs}, (err) => {
            (err === null).should.be.equal(true);
            enFs.link.callCount.should.be.equal(1);
            enFs.readFileSync(nodePath.join(dst, "subfolder1", "file3"), "utf8").should.containEql("data file 3");
            tearDownMockFs("EXDEV");
            done();
        });
    });
    //describe("> when trying to a move a folder into itself", function () {
    it("should produce an error", function (done) {
        const src = nodePath.join(tmpPath, "test");
        const dst = nodePath.join(tmpPath, "test", "test");
        (function () {
            enFs.statSync(src);
        }).should.throw(Error);
        enfsmkdirp.mkdirpSync(src);
        move(src, dst, (err) => {
            err.should.be.instanceOf(Error);
            enFs.statSync(src).isDirectory().should.be.equal(true);
            done();
        });
    });
    it("should overwrite the destination", function (done) {
        const src = nodePath.join(tmpPath, "src");
        const dst = tmpHelperPath;
        // use fixtures dir as dst since it has stuff
        const files = enFs.readdirSync(dst);
        files.indexOf("file1").should.be.greaterThanOrEqual(0);
        files.indexOf("folder1").should.be.greaterThanOrEqual(0);
        //create new src dir
        enfsmkdirp.mkdirpSync(nodePath.join(src, "folder-test"));
        enFs.writeFileSync(nodePath.join(src, "some-file"), "data some-file");
        const filesSrc = enFs.readdirSync(src);
        filesSrc.indexOf("some-file").should.be.greaterThanOrEqual(0);
        filesSrc.indexOf("folder-test").should.be.greaterThanOrEqual(0);
        move(src, dst, {overwrite: true, fs: enFs}, function (err) {
            (err === null).should.be.equal(true);
            const filesDst = enFs.readdirSync(dst);
            //dst should not have old stuff
            filesDst.indexOf("file1").should.be.equal(-1);
            filesDst.indexOf("folder1").should.be.equal(-1);
            //dst should have new stuff
            filesDst.indexOf("some-file").should.be.greaterThanOrEqual(0);
            filesDst.indexOf("folder-test").should.be.greaterThanOrEqual(0);
            done();
        });
    });
});


lib/moveSync.js
/**
 * @project enfsmove
 * @filename moveSync.js
 * @description sync method to move items on the file system
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */
"use strict";


const nodePath = require("path");
const enFs = require("enfspatch");
const enfsmkdirp = require("enfsmkdirp").mkdirpSync;
const enfscopy = require("enfscopy").copySync;
const rimraf = require("rimraf");



function moveAcrossDevice(src, dst, options) {
    const stats = options.fs.statSync(src);
    if (stats.isDirectory()) {
        return moveDirAcrossDevice(src, dst, options);
    } else {
        return moveFileAcrossDevice(src, dst, options);
    }
}


function moveFileAcrossDevice(src, dst, options) {
    try {
        enfscopy(src, dst);
    } catch (err) {
        options.fs.unlinkSync(dst);
        if (err.code === "EISDIR" || err.code === "EPERM") {
            return moveDirAcrossDevice(src, dst, options);
        } else {
            throw err;
        }
    }
}


function moveDirAcrossDevice(src, dst, options) {
    if (options.overwrite) {
        rimraf.sync(dst);
        enfscopy(src, dst, {fs: options.fs, stopOnError: true, overwrite: false, limit: options.limit});
    } else {
        enfscopy(src, dst, {fs: options.fs, stopOnError: true, overwrite: false, limit: options.limit});
    }
    rimraf.sync(src);
}

function doRename(src, dst, options) {
    if (options.overwrite) {
        try {
            return options.fs.renameSync(src, dst);
        } catch (err) {
            if (err.code === "ENOTEMPTY" || err.code === "EEXIST" || err.code === "EPERM") {
                rimraf.sync(dst);
                options.overwrite = false;
                return move(src, dst, options);
            }
            if (err.code === "EXDEV" || err.code === "EISDIR" || err.code === "EPERM") {
                return moveAcrossDevice(src, dst, options);
            }
        }
    } else {
        try {
            options.fs.linkSync(src, dst);
        } catch (err) {
            if (err.code === "EXDEV" || err.code === "EISDIR" || err.code === "EPERM") {
                return moveAcrossDevice(src, dst, options);
            }
            throw err;
        }
        options.fs.unlinkSync(src);
    }
}

function mkdirs(src, dst, options) {
    options.fs.mkdirpSync(nodePath.dirname(dst));
    doRename(src, dst, options);
}

/**
 * Move items in the file system sync
 * @param {string} src - the path to the items being moved
 * @param {string} dst - the destination path to where the items will be moved
 * @param {object} opt - various options for move module
 *              {object} fs - the fs module to be used
 *              {bool} mkdirp - if true will create new directories instead of copying the old ones
 *              {bool} overwrite - if true will overwrite items at destination if they exist
 *              {Number} limit - the maximum number of items being moved at a moment
 * @return {Error}
 */
function move(src, dst, opt) {
    const options = opt || {};
    options.fs = options.fs || enFs;
    options.fs.mkdirpSync = options.fs.mkdirpSync || enfsmkdirp;

    options.mkdirp = options.mkdirp !== false;

    options.overwrite = options.overwrite === true;
    options.limit = options.limit || 16;

    if (options.mkdirp) {
        mkdirs(src, dst, options);
    } else {
        doRename(src, dst, options);
    }
}

module.exports = move;


lib/moveAsync.js
/**
 * @project enfsmove
 * @filename moveAsync.js
 * @description async method to move items on the file system
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */
"use strict";


const nodePath = require("path");
const enFs = require("enfspatch");
const enfsmkdirp = require("enfsmkdirp").mkdirp;
const enfscopy = require("enfscopy").copy;
const rimraf = require("rimraf");

function noop() {
}

const kindOf = (arg) => arg === null ? "null" : typeof arg === "undefined" ? "undefined" : /^\[object (.*)\]$/.exec(Object.prototype.toString.call(arg))[1].toLowerCase();
const isFunction = (arg) => "function" === kindOf(arg);


function moveAcrossDevice(src, dst, options, callback) {
    options.fs.stat(src, (err, stat) => {
        if (err) {
            return callback(err);
        }
        if (stat.isDirectory()) {
            moveDirAcrossDevice(src, dst, options, callback);
        } else {
            moveFileAcrossDevice(src, dst, options, callback);
        }
    });
}


function moveFileAcrossDevice(src, dst, options, callback) {
    const outFlags = options.overwrite ? "w" : "wx";
    const ins = options.fs.createReadStream(src);
    const outs = options.fs.createWriteStream(dst, {flags: outFlags});

    const onClose = () => options.fs.unlink(src, callback);


    ins.on("error", (err) => {
        ins.destroy();
        outs.destroy();
        outs.removeListener("close", onClose);

        // don't care about error here
        options.fs.unlink(dst, () => {
            // note: `err` here is from the input stream errror
            if (err.code === "EISDIR" || err.code === "EPERM") {
                moveDirAcrossDevice(src, dst, options, callback);
            } else {
                callback(err);
            }
        });
    });

    outs.on("error", (err) => {
        ins.destroy();
        outs.destroy();
        outs.removeListener("close", onClose);
        callback(err);
    });

    outs.once("close", onClose);
    ins.pipe(outs);
}


function moveDirAcrossDevice(src, dst, options, callback) {

    function copyDir() {
        enfscopy(src, dst, {
            fs: options.fs,
            stopOnError: true,
            overwrite: false,
            limit: options.limit
        }, (errCopy) => {
            if (errCopy) {
                return callback(errCopy);
            }
            rimraf(src, callback);
        });
    }

    if (options.overwrite) {
        rimraf(dst, (err) => {
            if (err) {
                return callback(err);
            }
            copyDir();
        });
    } else {
        copyDir();
    }
}


function doRename(src, dst, options, callback) {
    if (options.overwrite) {
        options.fs.rename(src, dst, (err) => {
            if (!err) {
                return callback(null);
            }
            if (err.code === "ENOTEMPTY" || err.code === "EEXIST") {
                return rimraf(dst, (errRimRaf) => {
                    if (errRimRaf) {
                        return callback(errRimRaf);
                    }
                    options.overwrite = false;
                    move(src, dst, options, callback);
                });
            }
            // weird Windows shit
            if (err.code === "EPERM") {
                return setTimeout(() => {
                    rimraf(dst, (errRimRaf) => {
                        if (errRimRaf) {
                            return callback(errRimRaf);
                        }
                        options.overwrite = false;
                        move(src, dst, options, callback);
                    });
                }, 200);
            }
            if (err.code === "EXDEV" || err.code === "EISDIR" || err.code === "EPERM") {
                moveAcrossDevice(src, dst, options, callback);
            }
        });
    } else {
        options.fs.link(src, dst, (err) => {
            if (err) {
                if (err.code === "EXDEV" || err.code === "EISDIR" || err.code === "EPERM") {
                    return moveAcrossDevice(src, dst, options, callback);
                }
                return callback(err);
            }
            options.fs.unlink(src, callback);
        });
    }
}

function mkdirs(src, dst, options, callback) {
    options.fs.mkdirp(nodePath.dirname(dst), (err) => {
        if (err) {
            return callback(err);
        }
        doRename(src, dst, options, callback);
    });
}


/**
 * Move items in the file system async
 * @param {string} src - the path to the items being moved
 * @param {string} dst - the destination path to where the items will be moved
 * @param {object} opt - various options for move module
 *              {object} fs - the fs module to be used
 *              {bool} mkdirp - if true will create new directories instead of copying the old ones
 *              {bool} overwrite - if true will overwrite items at destination if they exist
 *              {Number} limit - the maximum number of items being moved at a moment
 * @param {function} callback - the callback function that will be called after the list is done
 * @return {Error}
 */
function move(src, dst, opt, callback) {
    if (isFunction(opt)) {
        callback = opt;
        opt = {};
    }
    callback = callback || noop;
    const options = opt || {};
    options.fs = options.fs || enFs;
    options.fs.mkdirp = options.fs.mkdirp || enfsmkdirp;

    options.mkdirp = options.mkdirp !== false;

    options.overwrite = options.overwrite === true;
    options.limit = options.limit || 512;

    if (options.mkdirp) {
        mkdirs(src, dst, options, callback);
    } else {
        doRename(src, dst, options, callback);
    }
}


module.exports = move;
